lib/js_routes.rb

```diff
@@ -216,13 +216,19 @@ def build_js(route, parent_route)
   end
 
   def route_js_arguments(route, parent_spec)
+    parts = route.parts
     required_parts = route.required_parts.clone
-    optional_parts = route.parts - required_parts
-    default_parts = route.defaults.select do |part, _|
+    optional_parts = parts - required_parts
+    default_options = route.defaults.select do |part, _|
       FILTERED_DEFAULT_PARTS.exclude?(part) && URL_OPTIONS.include?(part) || required_parts.include?(part)
     end
     [
-      required_parts, optional_parts, serialize(route.path.spec, parent_spec), default_parts
+      parts,
+      required_parts,
+      optional_parts,
+      default_options,
+      Rails.version < '4.0',
+      serialize(route.path.spec, parent_spec)
```

Comment:
@bogdan Hi, thanks for your feedback!

I don't agree about keeping only `required_parts` and `optional_parts` as it was before, third parameter was introduced for a reason.
It's impossible to merge these two arrays into original `parts` because the order of elements is unknown. Suppose something like `/(:opt1)/:arg1/(:opt2)/:arg2` — how do I know that `:arg1` comes before `:opt2` having only `[:opt1, :opt2]` and `[:arg1, :arg2]`? The order of elements in `parts` is crucial for correctness.

Instead, I propose to pass only `parts` and `required_parts` so I can restore `optional_parts` from it. Passing `parts` and `optional_parts` is also an option (in this case `required_parts` will be restored) but I assume that `required_parts` is usually shorter than `optional_parts` so I suggest to optimize with this assumption in the head.